palm/palm_config.py

```diff
@@ -39,6 +54,31 @@ def _get_config(self) -> object:
 
         return yaml.safe_load(config_path.read_text())
 
+    def _get_global_config(self, global_config_path: Optional[Path] = None) -> object:
+        """Gets the global config, reading yaml and returning a dict.
+        If the config does not exist, create it.
+
+        Args:
+            global_config_path (Optional[Path], optional): Used for testing. Defaults to None.
+
+        Returns:
+            object: dict of global config
+        """
+        config_path = global_config_path or Path().home() / '.palm' / 'config.yaml'
+        if not config_path.exists():
+            self._create_global_config_file(config_path)
+
+        return yaml.safe_load(config_path.read_text())
+
+    def _create_global_config_file(self, config_path) -> None:
+        """Creates the global config file."""
+        config_path.parent.mkdir(parents=True, exist_ok=True)
+        default_config = {
```

Comment:
Question is the reason that this line: `config_path.parent.mkdir(parents=True, exist_ok=True)` in _create_global_config_file() works because you defined the config_path variable in the _get_global_config()?

In other words, would the create config file function work if it was called within another function that had not already definied the config_path variable?

Reply:
config_path is a required argument to this function, if it's not provided there will be an exception raised.

I'm passing it into this function from the caller so that I don't have to define the variable twice.

In the calling function I'm doing something semi-strange, the optional arg there is purely so that I can test that function. There are other ways i could have handled that, like a class attribute which could be mocked in the tests. I just chose to take the path of least resistance, but that change could make the code a little cleaner (I'd be able to remove the `or` on line 67)

Reply:
yep @cgaunt-palmetto  this is a great example of "wait to refactor" IMO. The code is a little confusing (ie the weird dance at line 67 might be better as an attribute like Jake mentioned, or I personally might code golf this into a more succinct method), but that's because we don't really know how this code needs to play in the larger architecture yet since the global plugins feature is so new. So for me, I'd leave it exactly as it is and wait for it to be wrong ~ we could never touch this code again in 5 years, in which case who cares if it's confusing, or we could change it a bunch in the next 2 months, in which case we will know a lot more and be able to make smarter decisions about how we refactor ;)
![image](https://user-images.githubusercontent.com/82404678/154970935-d8de8642-f7b0-47cf-afef-98d077047f2c.png)